lib/spack/spack/version.py

```diff
@@ -402,7 +402,26 @@ def is_successor(self, other):
 
     @coerced
     def overlaps(self, other):
-        return self in other or other in self
+        """self.version = x.y
+           other.version = a.b.c
+           self overlaps other iff c==0
```

Comment:
could you please add a tests to lib/spack/spack/test/versions.py ?

Reply:
Ok that test opened more questions for me.

Let me state what I think the design principles involved here are:
* Patches should be applied if a spec version _overlaps_ a patch constraint not if a spec _satisfies_ a patch constraint. One can imagine the mpich patch situation happening with some regularity.

* Specs should for consistency with other package managers and ease of use the _satisfies_ semantics.  This will generally work as in a.b.c c's usually involve internal changes that can/will break patchs but do not change the public API.

Since overlap should have meaning different from satisfies the way the list ranges overlap seems counterintuitive based on the  defnition of version overlap as symetric and specific based on the conversion of x.y to the normal form of x.y.0.0.0...

this assertion fails:
```
assert_no_overlap('1.2.1', ':1.2')
```
It does this because
```
assert_in('1.4.2', '1.2:1.4')
```
is true.

__contains__ uses the satisfies semantics 😞 but overlap for ranges depends on it.

By my logic:
1.2.1 is in 1.2: and not in :1.2
while 1.2.1 satisifies both :1.2 and 1.2:

But I would propse the fix for patches to simply remove the dependency of overlaps on __contains__. 
Unfortunately even this breaks some other tests namely 
```
assert_overlaps('1.6.5', ':1.6')
test_intersection_with_containment
test_union_with_containment
```